### PR TITLE
[WIP] update go/loader to go/packages for module support

### PR DESCRIPTION
**Pull request by rdallman**
working on #50 

posting in case of bus incident to show mercy to poor other soul who may choose to spend their time attempting to accomplish this (there aren't really any docs for migrating, but managed to figure it out)... this hasn't begun to try to update refactoring package which is possibly a mine field, but I'm hoping not. the names analysis is working and passing tests, shouldn't take too long to plug in elsewhere to start testing. hopefully get some more time this week...

### CODE REVIEW

File: analysis/dataflow/dataflow_test.go [hunk 1]
```diff
@@ -522,55 +525,47 @@ func BenchmarkReaching(b *testing.B) {
...
-		b.FailNow()
-	}
+    `)
 
 	// create CFG and perform analyses
 	for n := 0; n < b.N; n++ {
-		cfg := cfg.FromFunc(f.Decls[0].(*ast.FuncDecl))
-		DefUse(cfg, prog.Created[0])
-		LiveVars(cfg, prog.Created[0])
+		cfg := cfg.FromFunc(c.prog.Syntax[0].Decls[0].(*ast.FuncDecl))
```
Comment (rdallman) [inline analysis/dataflow/dataflow_test.go:532]: note: everything that does c.prog.Syntax[0] is basically incorrect, many files are loaded, even if this is only used in benchmarking (I think?) so maybe doesn't matter

**Comment by sorenh** [comment]
Is this branch up-to-date? I'm curious to see the progress. :)

**Comment by sorenh** [comment]
It appears there's some work going into gopls to also support extraction to variable and function.

https://github.com/golang/tools/blob/master/internal/lsp/source/extract.go

I can't tell how much overlap there is, but wanted to point it out nonetheless.

**Comment by rdallman** [comment]
hey @sorenh - hit a bit of a snag, but have a plan and just need to find the time to do it, so I have a badly broken branch laying around for now. they changed the interaction with the filesystem quite a bit and I'm hoping to get it workable without having to drastically change the filesystem api we have (for now), we use this quite a bit for testing, modifying stdin, etc so it's relatively messy. been a busy couple of weeks for me.

interesting, that looks new. well, could cede to the overlords (that's the official package now), they appear to have the analysis needed to do it accurately but I haven't picked through it with a fine tooth comb which I guess could be useful, I hope they stole some stuff so that this was useful ;)

**Comment by sorenh** [comment]
Not sure if tagging like this works on Github, but @stamblerre was very responsive on some other work I did on gopls. Maybe reach out to make there's no duplication of effort.

**Comment by stamblerre** [comment]
Extracting functions and variables is now supported in `gopls` (https://github.com/golang/go/issues/37170). Are there other godoctor features that should be implemented in `gopls`?

**Comment by sorenh** [comment]
@stamblerre The vscode plugin still relies on godocter, doesn't it?

**Comment by stamblerre** [comment]
It does, but it will likely switch to `gopls` at some point in the near future. (cc @hyangah)

**Comment by rdallman** [comment]
it is relatively close, there are 4 tests failing (out of hundreds) so getting there. cgo and some extract func tests that are supposed to fail but don't. i know, it has been a year (and what a year)... but need to test actual support with modules after getting the tests all working, the tests were all before modules existed so likely need to make some modules tests and then (not in this patch) migrate the tests to all be modules compliant (our little GOPATH hackery has been taken from us)

**Comment by rdallman** [comment]
fixed the issue with build errors not flipping errors now, those tests work again (yay). snoozed and lost though and now actually have to migrate the tests to use modules if they import any packages, as the GOPATH hackery is no longer supported in 1.18 it seems like (I guess I could get them working on an older version but that seems like doubling my effort). still think cgo is broke too.

other thing i thought of is even if this does get back to working golang has made lots of changes to the lang we don't really have any test coverage for like generics, so we should maybe make a note of last working go version and use at your own peril warning - not sure this will get any love with gopls superceding, figured I'd try to get it working at least though...

**Comment by rdallman** [comment]
there is only one broken test now, for refactoring an inline C file (an imported cgo package seems fine). I managed to figure out that we can still do the GOPATH dance with GO111MODULES=off, which for some reason I had in my head was deprecated. so woo. 

from what I gather, the packages package (heh) actually does support loading cgo now as of 1.15 and the type checker or someone deep down seems to be creating a file in the cache with the contents of the file itself, but not the name we expect (some gobbledy gook name in the cache) so refactoring throws a fit since the file we want to refactor appears to not exist in the fileset. I'm down the rabbit hole trying to figure out how to get the file handle we want, it turns into the cache file in the token.FileSet and idk how to dig it back out. it's unclear how to simply disallow cgo too, there's a maze of configs it seems and nothing exposed from packages loader but maybe this is easy. 

so close!

**Comment by rdallman** [comment]
I feel kind of inclined to ship it since this will probably work with the majority of go programs now (vs master, which does not), and this is only a fringe case (inline cgo). but i'll sleep on it and think if there's a "nice" way to error at least.

**Comment by rdallman** [comment]
Considering master is broken against packages with modules, think this is a large improvement. *yells into void* :)

I did some testing on cross package refactors with and without modules and this seems to work now. I didn't spend too too much time on it but they seem to work out of the vim plugin, which is setting the scope, which is maybe something to update. 

re: the broken cgo case, the current broken test seems like it might be kinda broken functionality wise anyway per https://github.com/godoctor/godoctor/issues/22 - I didn't test this either.

File: analysis/loader/loader.go [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+	// Fset contains the entire program Fset
+	Fset *token.FileSet
+
+	// AllPackages is the list of all loaded packages for a program.
+	// It is this exposed this way to match the old loader API.
+	AllPackages map[*types.Package]*packages.Package
+}
+
+// Load loads a package, calling packages.Load
+func Load(conf *packages.Config, errorH func(error), args ...string) (*Program, error) {
```
Comment (QuLogic) [inline analysis/loader/loader.go:34]: All of the tests that were touched here now pass in `conf`, but none of them were updated to pass `errorH`.

Comment (rdallman) [reply analysis/loader/loader.go:34]: yea, good eye. the errorH function was basically to handle cgo somewhat more gracefully by catching it and spitting out a legible error. the new modules handles cgo better than the loader program so I believe it's dead code, though the broken test seems to be a bug maybe in modules handling (I gave up at the time). thanks for catching, should prob remove if that's the case

**Pull request merged**